File: ios-native/RemittanceApp/Services/CardScanner/CardScannerManager.swift
//
//  CardScannerManager.swift
//  RemittanceApp
//
//  Card scanning with OCR using Vision framework
//

import Foundation
import UIKit
import Vision
import AVFoundation

/// Card information extracted from scanning
struct ScannedCardInfo {
    var cardNumber: String?
    var expiryDate: String?
    var cardholderName: String?
    var cvv: String?
    var confidence: Float
}

/// Card scanner manager using Vision framework
class CardScannerManager: NSObject {
    
    // MARK: - Properties
    
    static let shared = CardScannerManager()
    
    private var captureSession: AVCaptureSession?
    private var previewLayer: AVCaptureVideoPreviewLayer?
    private var videoOutput: AVCaptureVideoDataOutput?
    
    private let sessionQueue = DispatchQueue(label: "com.remittance.cardscanner")
    private let visionQueue = DispatchQueue(label: "com.remittance.vision")
    
    private var isScanning = false
    private var scanCompletion: ((Result<ScannedCardInfo, Error>) -> Void)?
    
    // Card number regex patterns
    private let cardNumberPattern = #"(\d{4}[\s\-]?\d{4}[\s\-]?\d{4}[\s\-]?\d{4})"#
    private let expiryPattern = #"(0[1-9]|1[0-2])[\/\-](\d{2}|\d{4})"#
    private let cvvPattern = #"\b\d{3,4}\b"#
    
    // MARK: - Initialization
    
    private override init() {
        super.init()
    }
    
    // MARK: - Public Methods
    
    /// Check if device supports card scanning
    func isCardScanningSupported() -> Bool {
        return AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back) != nil
    }
    
    /// Request camera permission
    func requestCameraPermission(completion: @escaping (Bool) -> Void) {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            completion(true)
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { granted in
                DispatchQueue.main.async {
                    completion(granted)
                }
            }
        case .denied, .restricted:
            completion(false)
        @unknown default:
            completion(false)
        }
    }
    
    /// Setup camera session
    func setupCameraSession(previewView: UIView) throws -> AVCaptureVideoPreviewLayer {
        guard let device = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back) else {
            throw CardScannerError.cameraNotAvailable
        }
        
        let session = AVCaptureSession()
        session.sessionPreset = .high
        
        let input = try AVCaptureDeviceInput(device: device)
        guard session.canAddInput(input) else {
            throw CardScannerError.cannotAddInput
        }
        session.addInput(input)
        
        let output = AVCaptureVideoDataOutput()
        output.setSampleBufferDelegate(self, queue: visionQueue)
        output.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]
        
        guard session.canAddOutput(output) else {
            throw CardScannerError.cannotAddOutput
        }
        session.addOutput(output)
        
        self.captureSession = session
        self.videoOutput = output
        
        let previewLayer = AVCaptureVideoPreviewLayer(session: session)
        previewLayer.videoGravity = .resizeAspectFill
        previewLayer.frame = previewView.bounds
        
        self.previewLayer = previewLayer
        
        return previewLayer
    }
    
    /// Start scanning for card
    func startScanning(completion: @escaping (Result<ScannedCardInfo, Error>) -> Void) {
        guard let session = captureSession else {
            completion(.failure(CardScannerError.sessionNotSetup))
            return
        }
        
        self.scanCompletion = completion
        self.isScanning = true
        
        sessionQueue.async {
            session.startRunning()
        }
    }
    
    /// Stop scanning
    func stopScanning() {
        isScanning = false
        sessionQueue.async { [weak self] in
            self?.captureSession?.stopRunning()
        }
    }
    
    /// Scan image directly (for photo library images)
    func scanImage(_ image: UIImage, completion: @escaping (Result<ScannedCardInfo, Error>) -> Void) {
        guard let cgImage = image.cgImage else {
            completion(.failure(CardScannerError.invalidImage))
            return
        }
        
        let request = VNRecognizeTextRequest { [weak self] request, error in
            guard let self = self else { return }
            
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let observations = request.results as? [VNRecognizedTextObservation] else {
                completion(.failure(CardScannerError.noTextFound))
                return
            }
            
            let cardInfo = self.extractCardInfo(from: observations)
            
            if cardInfo.cardNumber != nil {
                completion(.success(cardInfo))
            } else {
                completion(.failure(CardScannerError.noCardDetected))
            }
        }
        
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = false
        
        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        
        visionQueue.async {
            do {
                try handler.perform([request])
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    // MARK: - Private Methods
    
    private func processVideoFrame(_ sampleBuffer: CMSampleBuffer) {
        guard isScanning else { return }
        
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            return
        }
        
        let request = VNRecognizeTextRequest { [weak self] request, error in
            guard let self = self else { return }
            
            if let error = error {
                print("Vision error: \(error.localizedDescription)")
                return
            }
            
            guard let observations = request.results as? [VNRecognizedTextObservation] else {
                return
            }
            
            let cardInfo = self.extractCardInfo(from: observations)
            
            // Only return if we have high confidence card number
            if let cardNumber = cardInfo.cardNumber, cardInfo.confidence > 0.7 {
                self.isScanning = false
                
                DispatchQueue.main.async {
                    self.scanCompletion?(.success(cardInfo))
                    self.stopScanning()
                }
            }
        }
        
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = false
        
        let handler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, options: [:])
        
        do {
            try handler.perform([request])
        } catch {
            print("Failed to perform Vision request: \(error)")
        }
    }
    
    private func extractCardInfo(from observations: [VNRecognizedTextObservation]) -> ScannedCardInfo {
        var cardInfo = ScannedCardInfo(confidence: 0)
        var allText: [String] = []
        var totalConfidence: Float = 0
        
        for observation in observations {
            guard let candidate = observation.topCandidates(1).first else { continue }
            allText.append(candidate.string)
            totalConfidence += candidate.confidence
        }
        
        cardInfo.confidence = observations.isEmpty ? 0 : totalConfidence / Float(observations.count)
        
        // Extract card number
        cardInfo.cardNumber = extractCardNumber(from: allText)
        
        // Extract expiry date
        cardInfo.expiryDate = extractExpiryDate(from: allText)
        
        // Extract cardholder name
        cardInfo.cardholderName = extractCardholderName(from: allText)
        
        return cardInfo
    }
    
    private func extractCardNumber(from texts: [String]) -> String? {
        let regex = try? NSRegularExpression(pattern: cardNumberPattern, options: [])
        
        for text in texts {
            let range = NSRange(text.startIndex..., in: text)
            if let match = regex?.firstMatch(in: text, options: [], range: range) {
                let matchedString = (text as NSString).substring(with: match.range)
                let cleaned = matchedString.replacingOccurrences(of: " ", with: "")
                    .replacingOccurrences(of: "-", with: "")
                
                // Validate using Luhn algorithm
                if isValidCardNumber(cleaned) {
                    return formatCardNumber(cleaned)
                }
            }
        }
        
        return nil
    }
    
    private func extractExpiryDate(from texts: [String]) -> String? {
        let regex = try? NSRegularExpression(pattern: expiryPattern, options: [])
        
        for text in texts {
            let range = NSRange(text.startIndex..., in: text)
            if let match = regex?.firstMatch(in: text, options: [], range: range) {
                let matchedString = (text as NSString).substring(with: match.range)
                return formatExpiryDate(matchedString)
            }
        }
        
        return nil
    }
    
    private func extractCardholderName(from texts: [String]) -> String? {
        // Look for text that appears to be a name (2-4 words, mostly letters)
        let namePattern = #"^[A-Z][A-Z\s]{5,30}$"#
        let regex = try? NSRegularExpression(pattern: namePattern, options: [])
        
        for text in texts {
            let upperText = text.uppercased()
            let range = NSRange(upperText.startIndex..., in: upperText)
            
            if regex?.firstMatch(in: upperText, options: [], range: range) != nil {
                // Exclude common card-related words
                let excludedWords = ["DEBIT", "CREDIT", "CARD", "BANK", "VALID", "THRU", "EXPIRES"]
                let containsExcluded = excludedWords.contains { upperText.contains($0) }
                
                if !containsExcluded {
                    return upperText
                }
            }
        }
        
        return nil
    }
    
    // MARK: - Validation Helpers
    
    private func isValidCardNumber(_ number: String) -> Bool {
        guard number.count >= 13 && number.count <= 19 else { return false }
        guard number.allSatisfy({ $0.isNumber }) else { return false }
        
        // Luhn algorithm
        let digits = number.compactMap { Int(String($0)) }
        var sum = 0
        var isSecond = false
        
        for digit in digits.reversed() {
            var current = digit
            if isSecond {
                current *= 2
                if current > 9 {
                    current -= 9
                }
            }
            sum += current
            isSecond.toggle()
        }
        
        return sum % 10 == 0
    }
    
    private func formatCardNumber(_ number: String) -> String {
        // Format as XXXX XXXX XXXX XXXX
        var formatted = ""
        for (index, char) in number.enumerated() {
            if index > 0 && index % 4 == 0 {
                formatted += " "
            }
            formatted.append(char)
        }
        return formatted
    }
    
    private func formatExpiryDate(_ date: String) -> String {
        // Format as MM/YY
        let cleaned = date.replacingOccurrences(of: "/", with: "")
            .replacingOccurrences(of: "-", with: "")
        
        if cleaned.count >= 4 {
            let month = String(cleaned.prefix(2))
            let year = String(cleaned.suffix(2))
            return "\(month)/\(year)"
        }
        
        return date
    }
    
    /// Get card type from number
    func getCardType(from cardNumber: String) -> CardType {
        let cleaned = cardNumber.replacingOccurrences(of: " ", with: "")
        
        if cleaned.hasPrefix("4") {
            return .visa
        } else if cleaned.hasPrefix("5") {
            return .mastercard
        } else if cleaned.hasPrefix("3") {
            return .amex
        } else if cleaned.hasPrefix("6") {
            return .discover
        }
        
        return .unknown
    }
}

// MARK: - AVCaptureVideoDataOutputSampleBufferDelegate

extension CardScannerManager: AVCaptureVideoDataOutputSampleBufferDelegate {
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        processVideoFrame(sampleBuffer)
    }
}

// MARK: - Supporting Types

enum CardType: String {
    case visa = "Visa"
    case mastercard = "Mastercard"
    case amex = "American Express"
    case discover = "Discover"
    case unknown = "Unknown"
}

enum CardScannerError: LocalizedError {
    case cameraNotAvailable
    case cannotAddInput
    case cannotAddOutput
    case sessionNotSetup
    case invalidImage
    case noTextFound
    case noCardDetected
    case permissionDenied
    
    var errorDescription: String? {
        switch self {
        case .cameraNotAvailable:
            return "Camera is not available on this device"
        case .cannotAddInput:
            return "Cannot add camera input to session"
        case .cannotAddOutput:
            return "Cannot add video output to session"
        case .sessionNotSetup:
            return "Camera session is not setup"
        case .invalidImage:
            return "Invalid image provided"
        case .noTextFound:
            return "No text found in image"
        case .noCardDetected:
            return "No valid card detected"
        case .permissionDenied:
            return "Camera permission denied"
        }
    }
}

File: android-native/app/src/main/java/com/remittance/app/scanner/CardScannerManager.kt
package com.pos54link.app.scanner

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

/**
 * Card information extracted from scanning
 */
data class ScannedCardInfo(
    val cardNumber: String? = null,
    val expiryDate: String? = null,
    val cardholderName: String? = null,
    val cvv: String? = null,
    val confidence: Float = 0f,
    val cardType: CardType = CardType.UNKNOWN
)

/**
 * Card types
 */
enum class CardType(val displayName: String) {
    VISA("Visa"),
    MASTERCARD("Mastercard"),
    AMEX("American Express"),
    DISCOVER("Discover"),
    UNKNOWN("Unknown")
}

/**
 * Card scanner manager using ML Kit Text Recognition
 */
class CardScannerManager(private val context: Context) {

    private val textRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
    private var cameraExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    private var imageAnalyzer: ImageAnalysis? = null
    private var camera: Camera? = null
    
    private var isScanning = false
    private var scanCallback: ((Result<ScannedCardInfo>) -> Unit)? = null
    
    // Regex patterns
    private val cardNumberPattern = Regex("""(\d{4}[\s\-]?\d{4}[\s\-]?\d{4}[\s\-]?\d{4})""")
    private val expiryPattern = Regex("""(0[1-9]|1[0-2])[\/\-](\d{2}|\d{4})""")
    private val cvvPattern = Regex("""\b\d{3,4}\b""")
    
    companion object {
        private const val REQUIRED_PERMISSION = Manifest.permission.CAMERA
        
        fun hasCameraPermission(context: Context): Boolean {
            return ContextCompat.checkSelfPermission(
                context,
                REQUIRED_PERMISSION
            ) == PackageManager.PERMISSION_GRANTED
        }
    }
    
    /**
     * Check if device supports card scanning
     */
    fun isCardScanningSupported(): Boolean {
        return context.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)
    }
    
    /**
     * Setup camera for card scanning
     */
    suspend fun setupCamera(
        lifecycleOwner: LifecycleOwner,
        previewView: PreviewView
    ): Result<Unit> = suspendCoroutine { continuation ->
        val cameraProviderFuture = ProcessCameraProvider.getInstance(context)
        
        cameraProviderFuture.addListener({
            try {
                val cameraProvider = cameraProviderFuture.get()
                
                // Preview
                val preview = Preview.Builder()
                    .build()
                    .also {
                        it.setSurfaceProvider(previewView.surfaceProvider)
                    }
                
                // Image analyzer
                imageAnalyzer = ImageAnalysis.Builder()
                    .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                    .build()
                    .also {
                        it.setAnalyzer(cameraExecutor) { imageProxy ->
                            processImageProxy(imageProxy)
                        }
                    }
                
                // Select back camera
                val cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA
                
                try {
                    // Unbind all use cases before rebinding
                    cameraProvider.unbindAll()
                    
                    // Bind use cases to camera
                    camera = cameraProvider.bindToLifecycle(
                        lifecycleOwner,
                        cameraSelector,
                        preview,
                        imageAnalyzer
                    )
                    
                    continuation.resume(Result.success(Unit))
                } catch (e: Exception) {
                    continuation.resume(Result.failure(CardScannerException.CameraBindingFailed(e)))
                }
                
            } catch (e: Exception) {
                continuation.resume(Result.failure(CardScannerException.CameraSetupFailed(e)))
            }
        }, ContextCompat.getMainExecutor(context))
    }
    
    /**
     * Start scanning for card
     */
    fun startScanning(callback: (Result<ScannedCardInfo>) -> Unit) {
        isScanning = true
        scanCallback = callback
    }
    
    /**
     * Stop scanning
     */
    fun stopScanning() {
        isScanning = false
        scanCallback = null
    }
    
    /**
     * Scan image directly (for gallery images)
     */
    suspend fun scanImage(bitmap: Bitmap): Result<ScannedCardInfo> = withContext(Dispatchers.IO) {
        suspendCoroutine { continuation ->
            val image = InputImage.fromBitmap(bitmap, 0)
            
            textRecognizer.process(image)
                .addOnSuccessListener { visionText ->
                    val cardInfo = extractCardInfo(visionText.text)
                    
                    if (cardInfo.cardNumber != null) {
                        continuation.resume(Result.success(cardInfo))
                    } else {
                        continuation.resume(
                            Result.failure(CardScannerException.NoCardDetected)
                        )
                    }
                }
                .addOnFailureListener { e ->
                    continuation.resume(Result.failure(e))
                }
        }
    }
    
    /**
     * Process camera image proxy
     */
    @androidx.camera.core.ExperimentalGetImage
    private fun processImageProxy(imageProxy: ImageProxy) {
        if (!isScanning) {
            imageProxy.close()
            return
        }
        
        val mediaImage = imageProxy.image
        if (mediaImage != null) {
            val image = InputImage.fromMediaImage(
                mediaImage,
                imageProxy.imageInfo.rotationDegrees
            )
            
            textRecognizer.process(image)
                .addOnSuccessListener { visionText ->
                    val cardInfo = extractCardInfo(visionText.text)
                    
                    // Only return if we have high confidence card number
                    if (cardInfo.cardNumber != null && cardInfo.confidence > 0.7f) {
                        isScanning = false
                        scanCallback?.invoke(Result.success(cardInfo))
                    }
                }
                .addOnFailureListener { e ->
                    // Continue scanning on failure
                }
                .addOnCompleteListener {
                    imageProxy.close()
                }
        } else {
            imageProxy.close()
        }
    }
    
    /**
     * Extract card information from recognized text
     */
    private fun extractCardInfo(text: String): ScannedCardInfo {
        val lines = text.split("\n")
        
        val cardNumber = extractCardNumber(lines)
        val expiryDate = extractExpiryDate(lines)
        val cardholderName = extractCardholderName(lines)
        val cardType = cardNumber?.let { getCardType(it) } ?: CardType.UNKNOWN
        
        // Calculate confidence based on what we found
        var confidence = 0f
        if (cardNumber != null) confidence += 0.5f
        if (expiryDate != null) confidence += 0.25f
        if (cardholderName != null) confidence += 0.25f
        
        return ScannedCardInfo(
            cardNumber = cardNumber,
            expiryDate = expiryDate,
            cardholderName = cardholderName,
            confidence = confidence,
            cardType = cardType
        )
    }
    
    /**
     * Extract card number from text lines
     */
    private fun extractCardNumber(lines: List<String>): String? {
        for (line in lines) {
            val match = cardNumberPattern.find(line)
            if (match != null) {
                val cleaned = match.value
                    .replace(" ", "")
                    .replace("-", "")
                
                if (isValidCardNumber(cleaned)) {
                    return formatCardNumber(cleaned)
                }
            }
        }
        return null
    }
    
    /**
     * Extract expiry date from text lines
     */
    private fun extractExpiryDate(lines: List<String>): String? {
        for (line in lines) {
            val match = expiryPattern.find(line)
            if (match != null) {
                return formatExpiryDate(match.value)
            }
        }
        return null
    }
    
    /**
     * Extract cardholder name from text lines
     */
    private fun extractCardholderName(lines: List<String>): String? {
        val namePattern = Regex("""^[A-Z][A-Z\s]{5,30}$""")
        val excludedWords = listOf("DEBIT", "CREDIT", "CARD", "BANK", "VALID", "THRU", "EXPIRES")
        
        for (line in lines) {
            val upperLine = line.uppercase()
            if (namePattern.matches(upperLine)) {
                val containsExcluded = excludedWords.any { upperLine.contains(it) }
                if (!containsExcluded) {
                    return upperLine
                }
            }
        }
        return null
    }
    
    /**
     * Validate card number using Luhn algorithm
     */
    private fun isValidCardNumber(number: String): Boolean {
        if (number.length < 13 || number.length > 19) return false
        if (!number.all { it.isDigit() }) return false
        
        var sum = 0
        var isSecond = false
        
        for (digit in number.reversed()) {
            var current = digit.toString().toInt()
            if (isSecond) {
                current *= 2
                if (current > 9) {
                    current -= 9
                }
            }
            sum += current
            isSecond = !isSecond
        }
        
        return sum % 10 == 0
    }
    
    /**
     * Format card number as XXXX XXXX XXXX XXXX
     */
    private fun formatCardNumber(number: String): String {
        return number.chunked(4).joinToString(" ")
    }
    
    /**
     * Format expiry date as MM/YY
     */
    private fun formatExpiryDate(date: String): String {
        val cleaned = date.replace("/", "").replace("-", "")
        
        return if (cleaned.length >= 4) {
            val month = cleaned.substring(0, 2)
            val year = cleaned.substring(cleaned.length - 2)
            "$month/$year"
        } else {
            date
        }
    }
    
    /**
     * Get card type from card number
     */
    fun getCardType(cardNumber: String): CardType {
        val cleaned = cardNumber.replace(" ", "")
        
        return when {
            cleaned.startsWith("4") -> CardType.VISA
            cleaned.startsWith("5") -> CardType.MASTERCARD
            cleaned.startsWith("3") -> CardType.AMEX
            cleaned.startsWith("6") -> CardType.DISCOVER
            else -> CardType.UNKNOWN
        }
    }
    
    /**
     * Release resources
     */
    fun release() {
        stopScanning()
        cameraExecutor.shutdown()
        textRecognizer.close()
    }
}

/**
 * Card scanner exceptions
 */
sealed class CardScannerException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    object CameraNotAvailable : CardScannerException("Camera is not available on this device")
    object PermissionDenied : CardScannerException("Camera permission denied")
    object NoCardDetected : CardScannerException("No valid card detected")
    data class CameraSetupFailed(val cause: Throwable) : CardScannerException("Camera setup failed", cause)
    data class CameraBindingFailed(val cause: Throwable) : CardScannerException("Camera binding failed", cause)
}
